llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -612,6 +612,118 @@ Instruction *emitSpecConstantRecursive(Type *Ty, Instruction *InsertBefore,
 
 } // namespace
 
+/// Function creates load instruction from the given Buffer by the given Offset.
+/// Function returns the value of load instruction.
+static Value *createLoadFromBuffer(CallInst *CI, Value *Buffer, size_t Offset,
+                                   Type *SCType) {
+  LLVMContext &C = CI->getContext();
+  Type *Int8Ty = Type::getInt8Ty(C);
+  Type *Int32Ty = Type::getInt32Ty(C);
+  GetElementPtrInst *GEP = GetElementPtrInst::Create(
+      Int8Ty, Buffer, {ConstantInt::get(Int32Ty, Offset, false)}, "gep", CI);
+
+  Instruction *BitCast = nullptr;
+  if (SCType->isIntegerTy(1)) // No bitcast to i1 before load
+    BitCast = GEP;
+  else
+    BitCast = new BitCastInst(
+        GEP, PointerType::get(SCType, GEP->getAddressSpace()), "bc", CI);
+
+  // When we encounter i1 spec constant, we still load the whole byte
+  Value *Load = new LoadInst(SCType->isIntegerTy(1) ? Int8Ty : SCType, BitCast,
+                             "load", CI);
+  if (SCType->isIntegerTy(1)) // trunc back to i1 if necessary
+    Load = CastInst::CreateIntegerCast(Load, SCType, /* IsSigned */ false,
+                                       "tobool", CI);
+
+  return Load;
+}
+
+/// Function tries to dig out the initializer from the given CallInst to
+/// SpecConst function. ArgIndex is the expected index of the function operand
+/// leading to the initializer.
+///
+/// Examples:
+/// 1)
+///   %"spec_id" = type { i32 }
+///   @value = internal addrspace(1) constant %"spec_id" { i32 123 }, align 4
+///   call spir_func i32 @sycl_getScalar2020SpecConst(%1, @value, %2)
+///
+/// 2)
+///   %"spec_id" = type { %A }
+///   %A = type { i32 }
+///   @value = constant %"spec_id" { %A { i32 1 } }, align 4
+///   call spir_func void @getCompositeSpecConst(%1, %2, @value, %3)
+static Constant *tryGetSpecConstInitializerFromCI(CallInst *CI,
```

Comment:
With SYCL 2020, we must always have a default value available, which means that the function should not be called `try`, we should not use `dyn_cast<GlobalVariable>`, but instead use `cast`, etc. The latter could be a bit of a problem in case there is a complex chain of instructions leading from the argument to a global variable, but it should be compile-time resolvable with something like [`llvm::getUnderlyingObject`](https://llvm.org/doxygen/namespacellvm.html#a38eafb1665bac4e930d3b17491aab4a8), for example.

Once again, it is probably better to have this change in a separate PR, because it may cause some regressions if we miss some corner cases. Better not mix it with a non-functional refactoring, but let's have a `TODO` comment here somewhere about it.

Reply:
`With SYCL 2020, we must always have a default value available`

But here we check that sycl-post-link is able to come up with default value from type
https://github.com/intel/llvm/blob/b7a43a423b60d5e2e3ab6210307d7ec5b9fbaad0/llvm/test/tools/sycl-post-link/spec-constants/composite-default-value.ll#L4-L6

Reply:
That is likely an outdated/legacy test case and the comment. Yes, old 1.2.1 extension implementation used to set default values based on type in AOT mode, because there were no way to set them in source code. 

With SYCL 2020, default value is defined by user when [constructing `specialization_id`](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_constructors) object:

> Effects: Constructs a `specialization_id` containing an instance of `T` initialized with `args...`, which represents the specialization constant’s default value.

And the value is [guaranteed to be compile-time known](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_declaring_a_specialization_constant):

> the `specialization_id` variable must be declared as `constexpr`;

Reply:
I removed this test.